pymeasure/instruments/thorlabs/thorlabspm100usb.py

```diff
@@ -23,74 +23,97 @@
+    )
+
+    @property
+    def wavelength(self):
+        """Wavelength in nm."""
+        value = self.values("SENSE:CORR:WAV?")[0]
+        return value
+
+    @wavelength.setter
+    def wavelength(self, value):
+        """Wavelength in nm."""
+        if self.wavelength_settable:
+            value = truncated_range(value, [self.wavelength_min, self.wavelength_max])
+            self.write("SENSE:CORR:WAV {}".format(value))
+        else:
+            raise AttributeError(
+                f"{self.sensor_name} does not allow setting the wavelength."
+            )
+
+    @property
+    def power(self):
+        """Power in W."""
+        if self.is_power:
+            return self.values("MEAS:POW?")[0]
+        else:
+            raise AttributeError(f"{self.sensor_name} is not a power sensor.")
+
+    @property
+    def energy(self):
+        """Energy in J."""
+        if self.is_energy:
+            return self.values("MEAS:ENER?")[0]
+        else:
+            raise AttributeError(f"{self.sensor_name} is not an energy sensor.")
+
+    def _set_flags(self):
+        """Get sensor info and write flags."""
+        response = self.ask("SYST:SENSOR:IDN?").split(",")
         self.sensor_name = response[0]
         self.sensor_sn = response[1]
         self.sensor_cal_msg = response[2]
         self.sensor_type = response[3]
         self.sensor_subtype = response[4]
-        self._flags_str = response[-1][:-1]
-        # interpretation of the flags
-        # rough trick using bin repr, maybe something more elegant exixts
-        # (bitshift, bitarray?)
-        self._flags = tuple(
-            map(lambda x: x == '1', bin(int(self._flags_str))[2:]))
-        # setting the flags; _dn are empty
-        self.is_power, self.is_energy, _d4, _d8, \
-        self.resp_settable, self.wavelength_settable, self.tau_settable, _d128, self.temperature_sens = self._flags
+        _flags_str = response[5].rstrip("\n")
 
-    @property
-    def energy(self):
-        if self.is_energy:
-            return self.values("MEAS:ENER?")
-        else:
-            raise Exception("%s is not an energy sensor" % self.sensor_name)
-            return 0
+        # interpretation of the flags, see p. 49 of the manual:
+        # https://www.thorlabs.de/_sd.cfm?fileName=17654-D02.pdf&partNumber=PM100D
+
+        # Convert to binary representation and pad zeros to 9 bit for sensors
+        # where not all flags are present.
+        _flags_str = format(int(_flags_str), "09b")
+        # Reverse the order so it matches the flag order from the manual, i.e.
+        # from decimal values from 1 to 256.
+        _flags_str = _flags_str[::-1]
+
+        # Convert to boolean.
+        self.flags = [x == "1" for x in _flags_str]
 
-    @energy.setter
-    def energy(self, val):
-        raise Exception("Energy not settable!")
+        # setting the flags; _dn are unused; decimal values as comments
+        (
+            self.is_power,  # 1
```

Comment:
while we're improving the code, maybe `self.measures_power` and `self.measures_energy` are better names?

Reply:
I renamed all flags the names from the manual: c2e6e1f